edtlib: Match any parent bus when binding lacks an explicit on-bus

jhedberg opened the pull request:
There are some drivers in the tree that support devices on multiple
different buses, although so far this has not been represented in
device tree using the bus concept. In order to convert these drivers &
bindings to refer to a formal bus in device tree we need to be able to
match bindings which lack an explicit "on-bus: ..." value against any
parent bus.

The two drivers I'm particularly targeting is the ns16550 UART driver
(drivers/serial/uart_ns16550.c) and the DW I2C driver
(drivers/i2c/i2c_dw.c). They both support devices with a fixed MMIO
address as well as devices connected and discovered over PCIe. The
only issue is that instead of encoding the bus information the proper
DT way these bindings use a special "pcie" property in the DT node
entries to indicate whether the node is on the PCIe bus or not.

Being able to convert the above two drivers to use the DT bus concept
allow the removal of "hacks" like this:

```
 #if DT_INST_PROP(0, pcie) || \
       DT_INST_PROP(1, pcie) || \
       DT_INST_PROP(2, pcie) || \
       DT_INST_PROP(3, pcie)
```

to the more intuitive:

```
 #if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
```

This also has the benefit that the driver doesn't need to make any
arbitrary assumptions of how many matching devices there may be but
works for any number of matches. This is already a problem now since
e.g. the ns16550 driver assumes a maximum of 4 nodes, whereas
dts/x86/elkhart_lake.dtsi defines up to 9 different ns16550 nodes.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
edtlib: Match any parent bus when binding lacks an explicit on-bus

There are some drivers in the tree that support devices on multiple
different buses, although so far this has not been represented in
device tree using the bus concept. In order to convert these drivers &
bindings to refer to a formal bus in device tree we need to be able to
match bindings which lack an explicit "on-bus: ..." value against any
parent bus. This will also be needed for any external bindings, since
those would not be aware of on-bus (as it's a Zephyhr-specific
extension).

The two drivers I'm particularly targeting is the ns16550 UART driver
(drivers/serial/uart_ns16550.c) and the DW I2C driver
(drivers/i2c/i2c_dw.c). They both support devices with a fixed MMIO
address as well as devices connected and discovered over PCIe. The
only issue is that instead of encoding the bus information the proper
DT way these bindings use a special "pcie" property in the DT node
entries to indicate whether the node is on the PCIe bus or not.

Being able to convert the above two drivers to use the DT bus concept
allow the removal of "hacks" like this:

 if DT_INST_PROP(0, pcie) || \
       DT_INST_PROP(1, pcie) || \
       DT_INST_PROP(2, pcie) || \
       DT_INST_PROP(3, pcie)

to the more intuitive:

 if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)

This also has the benefit that the driver doesn't need to make any
arbitrary assumptions of how many matching devices there may be but
works for any number of matches. This is already a problem now since
e.g. the ns16550 driver assumes a maximum of 4 nodes, whereas
dts/x86/elkhart_lake.dtsi defines up to 9 different ns16550 nodes.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

galak commented:
I'd rather see us add proper PCI/PCIe support and represent the DTS nodes accordingly:

https://www.devicetree.org/open-firmware/bindings/pci/pci2_1.pdf

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/pci/pci.txt

(not sure if this binding applies to any of the intel SoCs we support in Zephyr:)
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/pci/intel-gw-pcie.yaml

jhedberg commented:
> I'd rather see us add proper PCI/PCIe support and represent the DTS nodes accordingly

@galak that's pretty much my plan, i.e. the next step is to define a binding for a PCIe bus and then reorganise all PCIe-based ns16550 nodes in device tree under parent PCIe bus nodes. Anyway, I guess that's kind of independent of this particular PR which just gives a way to define bindings that can match multiple different busses, which in turn allows accurately describing the status quo of our i2c_dw and ns16500 drivers in proper device tree terms.

The Linux specifications for PCI look quite big and complex, so at least initially I wasn't planning to try to implement all of them, but I will try to follow the structure and naming of the specs as much as possible.

galak commented on scripts/dts/edtlib.py:
```diff
@@ -773,6 +773,19 @@ def _init_binding(self):
+                for (compat, on_bus) in self.edt._compat2binding.keys():
+                    if on_bus is not None:
+                        continue
+                    if  compat in self.compats:
```
nit: 2 spaces after `if`

jhedberg replied:
fixed

mbolivar-nordic commented on scripts/dts/edtlib.py:
```diff
@@ -773,6 +773,19 @@ def _init_binding(self):
+                # If matching for the specific on_bus value failed look for
+                # bindings which don't have an explicitly defined bus (i.e.
+                # can support any bus).
+                for (compat, on_bus) in self.edt._compat2binding.keys():
+                    if on_bus is not None:
+                        continue
+                    if compat in self.compats:
+                        binding = self.edt._compat2binding[compat, on_bus]
+                        self.binding_path = binding.path
+                        self.matching_compat = compat
+                        self._binding = binding
+                        return
```
Why can't you just look up `self.edt._compat2binding[compat, None]`?

jhedberg replied:
@mbolivar-nordic I certainly can. It's just that my python-foo wasn't quite sharp enough so I didn't think about that option. It'd look like the following then, I suppose?
```
if (compat, None) in self.edt._compat2binding:
    binding = self.edt._compat2binding[compat, None]
    self.binding_path = binding.path
    self.matching_compat = compat
    self._binding = binding
    return
```

I take it that you're not in general opposed to this feature then?

jhedberg replied:
Actually, I guess I can't use None in the variable portion of the if-statement. So I should just directly look up from the dictionary and handle the case (exception?) that the key isn't there?

mbolivar-nordic replied:
>  It'd look like the following then, I suppose?

Seems right to me. A test case in testedtlib.py verifying this would also help.

> I take it that you're not in general opposed to this feature then?

Seems reasonable to me given the rest of the discussion here that it's a standalone generalization not directly related to PCIe.

jhedberg replied:
> Actually, I guess I can't use None in the variable portion of the if-statement. 

Hmm, I just attempted a build with that and didn't get any errors, so maybe that is correct syntax after all (i.e. None means that the portion of the key gets ignored and not assigned to any variable)

mbolivar-nordic replied:
I don't have any issues with this:

```python
d = {('vnd,foo', 'i2c'): 'binding1', ('vnd,foo', None): 'binding2'}
('vnd,foo', None) in d  # True
d['vnd,foo', None]  #  'binding2'
```

In general, `d[foo, bar]` is the same as `d[(foo, bar)]`, i.e. multiple arguments get wrapped up in a tuple as syntactic sugar, and the `_compat2binding` dict contains `(compatible, bus)` tuples.

mbolivar-nordic replied:
>  A test case in testedtlib.py

If you can add these, I think I'm ready to +1

jhedberg replied:
> If you can add these, I think I'm ready to +1

Test cases have been added.

jhedberg commented:
@mbolivar updated now, please take another look. Two main changes:

 - Moved the second check to a new for-loop so that we test for explicit bus values on all compats before attempting the "any bus" check
 - Added a `if on_bus is not None:` test before the second for-loop, since otherwise the first loop would already have gotten a match if one exists

mbolivar-nordic commented:
Thanks for the update.

> * Moved the second check to a new for-loop so that we test for explicit bus values on all compats before attempting the "any bus" check

So the question is basically whether a node with `compatible = "a", "b";` on an I2C bus should prefer:

1. a binding for `"b"` on an I2C bus, or
2. a binding for `"a"` on no bus at all

Assuming those are the only two bindings available, right?

I would have naively guessed that the tooling would match the most specific compatible first if possible, even after this change, but I really don't know what the 'right' thing is. Curious what motivated this change and what other people think.

jhedberg commented:
> I would have naively guessed that the tooling would match the most specific compatible first if possible

I think the thought was that a binding with an explicit `on-bus: ...` carries with it a stronger specificity than its position within the `compatible = ...` array. This would also maintain backwards compatible behaviour, i.e. if we take your example: if binding "b" was selected before this PR it will also be selected after this PR. However, I'm fairly new to these subtleties of device tree and haven't really done much deep thinking about it. I'm fine either way (mainly because I think this is unlikely to be anything but a theoretical concern) and can update the patch if we so decide.

mbolivar-nordic commented:
The compatibility argument is convincing to me.  @pabigot , @galak ?

galak commented:
> The compatibility argument is convincing to me. @pabigot , @galak ?

What case are you guys concerned about w/regards to `compatible` and `on-bus`?

jhedberg commented:
@galak I'm not sure I can explain it any better than what @mbolivar-nordic did in his earlier comment, but I'll try. Basically it's about the order in which entries in `compatible` are matched against bindings.

Given `compatible = "a", "b";` we had two options for how to do the matching order, either:

1. "a" exact bus match
2. "b" exact bus match
3. "a" any bus match
4. "b" any bus match

or:

1. "a" exact bus match
2. "a" any bus match
3. "b" exact bus match
4. "b" any bus match

And what we agreed upon (or at least what I've now implemented) is the first option.

pabigot commented:
Starting from first principles: As far as I can tell the whole "on-bus" thing is specific to Zephyr's binding description format.  There's nothing about buses anywhere in DT Schema that I can see, nor in Linux.

From that perspective I think "on-bus" should not be a primary consideration from the perspective of checking whether compatibles are supported.

(Then I have questions related to #29904 and #20669, which come down to: if a devicetree node has multiple compatibles, and the bindings search path has descriptions for more than one, the current rule is the binding for the earliest compatible in the list is picked, right?  What happens if the driver for that compatible is not enabled?  Or are symbols generated for all compatibles, so if the driver for the next compatible is enabled it would work?  Enquiring minds want to know, but perhaps not specifically relevant here.)

In any case my interpretation from the devicetree perspective is the first compatible that can be supported should be picked.  So that'd be the option not currently taken in https://github.com/zephyrproject-rtos/zephyr/pull/30928#issuecomment-749560237: `a` with bus; `a` without bus`, `b` with bus, `b` without bus.

This fits with my feeling that if some device is normally on a SPI or I2C bus, and I have a driver for a more specific compatible that is not on a bus, I want that driver to be used, whether or not the node is topographically a child of a bus.  A real-world example would be a binding for a SPI NOR flash device with:

    compatible = "vendor,qspi-flash", "jedec,spi-nor";

where the first doesn't have `on-bus` and the node is a child of a SPI bus.  Here if the QSPI flash driver is enabled that's what I want used in preference to the generic SPI NOR driver.

jhedberg commented:
@pabigot that argument about the ordering sounds quite reasonable to me, i.e. while I was attempting keep the behaviour compatible with the current Zephyr implementation, it might be even better to try to be compatible with generic (non-Zephyr specific) DT where bus definitions don't play any role in selecting the compatible.

At this point I'm happy to implement this either way, as long as you guys can agree on one way or the other :)

henrikbrixandersen commented:
I agree with @pabigot on this. No matter what we decide to go with, please incorporate a comment stating why this was chosen (either in code or documentation) so the reason isn't forgotten.

jhedberg commented:
Ok, so I also find @pabigot 's argument more convincing than my original one, so I've pushed an update which changes to the other order of matching. I also added justifications as requested by @henrikbrixandersen .

henrikbrixandersen commented on scripts/dts/test-bindings/device-on-any-bus.yaml:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: BSD-3-Clause
```
Is the license here correct?

jhedberg replied:
It’s a copy-paste of the other yaml files in this directory, but I’m also surprised to see that it’s not Apache 2. @mbolivar-nordic @galak ?

jhedberg replied:
@henrikbrixandersen upon closer inspection the entire Zephyr DT implementation (i.e. everything under `scripts/dts`) seems to be BSD 3-clause, so if you think there's a concern with that it needs to be addresses independently to this PR. With that in mind, do you want to reconsider your "changes requested"? :)

jhedberg commented:
@galak @mbolivar-nordic any further comments on this or is it ready to be merged?

galak commented on scripts/dts/edtlib.py:
```diff
@@ -768,13 +768,23 @@ def _init_binding(self):
+                # and for each one first try to match against an explicitly
+                # specified bus (if any) and then against any bus. This is so
+                # that matching against bindings which do not specify a bus
+                # (i.e. all Zephyr-external ones, since this is a Zephyr
```
nit: not sure what the (ie..) part of the comment is trying to convey.  As at some point I'm guessing in zephyr we'll have bindings that utilize this functionality.

jhedberg replied:
True. I think "e.g." would be more accurate. I can still change it if you want.

jhedberg replied:
I removed the entire comment and added a similar mention to the commit message
